httpclient5/src/main/java/org/apache/hc/client5/http/impl/auth/RFC5987Codec.java

```diff
@@ -0,0 +1,155 @@
+        for (int i = 'A'; i <= 'Z'; i++) {
+            UNRESERVED.set(i);
+        }
+        for (int i = '0'; i <= '9'; i++) {
+            UNRESERVED.set(i);
+        }
+
+        // Additional characters as per RFC 5987 attr-char
+        UNRESERVED.set('!');
+        UNRESERVED.set('#');
+        UNRESERVED.set('$');
+        UNRESERVED.set('&');
+        UNRESERVED.set('+');
+        UNRESERVED.set('-');
+        UNRESERVED.set('.');
+        UNRESERVED.set('^');
+        UNRESERVED.set('_');
+        UNRESERVED.set('`');
+        UNRESERVED.set('|');
+        UNRESERVED.set('~');
+    }
+
+
+
+    /**
+     * Encodes a string using the default UTF-8 charset.
+     *
+     * @param s The string to encode.
+     * @return The percent-encoded string.
+     */
+    static String encode(final String s) {
+        return encode(s, StandardCharsets.UTF_8);
+    }
+
+    /**
+     * Encodes a string using the specified charset.
+     *
+     * @param s       The string to encode.
+     * @param charset The charset to use for encoding.
+     * @return The percent-encoded string.
+     */
+    static String encode(final String s, final Charset charset) {
+        final ByteBuffer bb = charset.encode(CharBuffer.wrap(s));
+        final StringBuilder sb = new StringBuilder();
+
+        while (bb.hasRemaining()) {
+            final int b = bb.get() & 0xff;
+            if (UNRESERVED.get(b)) {
+                sb.append((char) b);
+            } else {
+                sb.append('%');
+                sb.append(Character.forDigit((b >> 4) & 0xF, RADIX));
+                sb.append(Character.forDigit(b & 0xF, RADIX));
+            }
+        }
+
+        return sb.toString();
+    }
+
+    /**
+     * Decodes a percent-encoded string using the default UTF-8 charset.
+     *
+     * @param s The percent-encoded string to decode.
+     * @return The decoded string.
+     */
+    static String decode(final String s) {
+        return decode(s, StandardCharsets.UTF_8);
+    }
+
+    /**
+     * Decodes a percent-encoded string using the specified charset.
+     *
+     * @param s       The percent-encoded string to decode.
+     * @param charset The charset to use for decoding.
+     * @return The decoded string.
+     * @throws IllegalArgumentException If the percent-encoded string is invalid.
+     */
+    static String decode(final String s, final Charset charset) {
+        final ByteBuffer bb = ByteBuffer.allocate(s.length());
+        final CharBuffer cb = CharBuffer.wrap(s);
+
+        while (cb.hasRemaining()) {
+            final char c = cb.get();
+            if (c == '%' && cb.remaining() >= 2) {
+                final int u = Character.digit(cb.get(), RADIX);
+                final int l = Character.digit(cb.get(), RADIX);
+                if (u != -1 && l != -1) {
+                    bb.put((byte) ((u << 4) + l));
+                } else {
+                    throw new IllegalArgumentException("Invalid percent encoding in " + s);
```

Comment:
@arturobernalg Can we throw a better exception here, like `CodingException` of some sort, which could be caught at the a higher level and translated in a proper protocol exception?